### PR TITLE
refactor autocast set/get funcs

### LINKED ISSUE #104478: AttributeError: module 'torch.mps' has no attribute 'is_autocast_enabled'
### 🐛 Describe the bug

with pytorch nightly, seeing this error during MPS usage

```
  File "torch/utils/checkpoint.py", line 192, in forward
    ctx.device_autocast_kwargs, ctx.cpu_autocast_kwargs = _get_autocast_kwargs(
  File "torch/utils/checkpoint.py", line 170, in _get_autocast_kwargs
    "enabled": device_module.is_autocast_enabled(),

AttributeError: module 'torch.mps' has no attribute 'is_autocast_enabled'
```

seems like it started with https://github.com/pytorch/pytorch/pull/99626

### Versions

PyTorch version: 2.1.0.dev20230630
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.4 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.25.2
Libc version: N/A

Python version: 3.10.12 (main, Jun  7 2023, 00:38:32) [Clang 14.0.3 (clang-1403.0.22.14.1)] (64-bit runtime)
Python platform: macOS-13.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Ultra

Versions of relevant libraries:
[pip3] numpy==1.25.0
[pip3] torch==2.1.0.dev20230630
[pip3] torchaudio==2.1.0.dev20230630
[pip3] torchvision==0.16.0.dev20230630
[conda] numpy                     1.21.5                   pypi_0    pypi
[conda] numpydoc                  1.5.0           py310hca03da5_0  
[conda] torch                     2.1.0.dev20230601          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230601          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes https://github.com/pytorch/pytorch/issues/104478
1. As for autocast (AMP), there are too many simliar APIs for different device, such as `torch.set_autocast_cpu_enabled` for `cpu`, `torch.set_autocast_ipu_enabled` for `ipu` and `xla`, `xpu` devices. If we want to support more device, for `mps`, we will add `torch.set_autocast_mps_enabled`, the front-end code in Python will be very complicated and messy. So I refactor these APIs and normalized them, then we can call a same API with device argument for different device.
2. And for `xla`, `ipu`, the related API is merged last week, so I removed it and it will not raise error for users. And for `cuda`, `cpu`, we add warning info to suggest to use the new APIs.
@albanD @soulitzer 

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

### CODE REVIEW

**Review by soulitzer — COMMENTED** [review]
Great clean up!
~It does looks bc-breaking though (we are removing `set_autocast_ipu_dtype`, etc).
Though it looks like nothing comes up from a quick github search, let's properly deprecate those as well instead of just deleting for now.~
Oh I didn't realize those were added recently.

File: torch/amp/autocast_mode.py [hunk 1]
```diff
@@ -228,53 +204,26 @@ def __init__(self, device_type : str,
...
+            "hpu": [torch.bfloat16, torch.float16],
+            "xla": [torch.bfloat16],
+        }
+        if self.custom_backend_name == self.device:
+            self.device_supported_fast_dtype_map[self.custom_backend_name] = \
+                getattr(torch, self.device).get_amp_supported_dtype()
+        if self.device == "cuda" and enabled and self.fast_dtype == torch.bfloat16 \
+                and not torch.cuda.is_bf16_supported():
+            raise RuntimeError('Current CUDA Device does not support bfloat16. Please switch dtype to float16.')
+        supported_dtyps = self.device_supported_fast_dtype_map[self.device]
```
Comment (soulitzer) [inline torch/amp/autocast_mode.py:221]: nit: supported_dtyps -> supported_dtypes

**Review by soulitzer — COMMENTED** [review]
Had some questions on the `__init__` part of autocast around custom backends, rest looks okay to me (just small nits).

**Comment by soulitzer** [comment]
cc @cowanmeg @charlie-wt for visibility

File: torch/amp/autocast_mode.py [hunk 2]
```diff
@@ -189,37 +192,10 @@ def __init__(self, device_type : str,
...
-        elif self.device == 'xpu':
-            self.fast_dtype = torch.xpu.get_autocast_xpu_dtype()  # type: ignore[attr-defined]
-        elif self.device == 'ipu':
-            self.fast_dtype = torch.get_autocast_ipu_dtype()  # type: ignore[attr-defined]
-        elif self.device == 'hpu':
-            self.fast_dtype = torch.hpu.get_autocast_hpu_dtype()  # type: ignore[attr-defined]
-        elif self.device == 'xla':
-            self.fast_dtype = torch.get_autocast_xla_dtype()  # type: ignore[attr-defined]
-        elif self.device == self.custom_backend_name:
-            necessary_funcs = ['is_autocast_enabled', 'set_autocast_enabled', 'get_autocast_dtype',
```
[the remark this answers is not in the document]
Comment (heidongxianhua) [reply torch/amp/autocast_mode.py:206]: > This changes what we expect custom backends to have to implement for AMP support right?
> 
> * Do previous custom backend module's implementations just end up calling into something like `at::autocast::set_privateuseone_enabled(enabled)`? One risk if the user defined `torch.foo.set_autocast_dtype` does some kind of side effect, that would break their code since it is no longer called.
> * It also looks like we still use `getattr(torch, self.device).get_amp_supported_dtype()` later in this PR, should we keep the proper error checking for that here?
> 
> This does look like it would clean up things, but I would be curious what @bdhirsh thinks.

yeah, I have add the `func` check and testcast for `get_amp_supported_dtype` defined in custom device backend. And as for the func `at::autocast::set_privateuseone_enabled(enabled)`, it is just defined in c++, and it does not export to Python, so I think it is ok. @soulitzer

File: torch/amp/autocast_mode.py (hunk 1 above)
Comment (heidongxianhua) [reply torch/amp/autocast_mode.py]: > nit: supported_dtyps -> supported_dtypes

could you have a look again? @soulitzer

File: torch/amp/autocast_mode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/amp/autocast_mode.py:206]: Going to defer to @bdhirsh on this one

**Review by soulitzer — CHANGES_REQUESTED** [review]
Thanks for the work on this PR and making the update, I think this is a great clean up and seems to be in the right direction. However, since this is a major change to the autocast public API, I think its worth stepping back and putting this in perspective to our plans with improving the autocast API surface in general. 

There's a lot of things going on that I think are worth addressing up front rather than having one-off PRs. For example, we have both `torch.cuda.amp.autocast(args...)` and `torch.autocast("cuda", args...)`. 
We also have a bunch of `torch.autocast_set_enabled` APIs directly on the torch name space, when we really want them to live in the amp namespace. They aren't really documented anywhere either. How do we expect users to use those low-level APIs or should they be private now? 
We also have the issue of custom backends. What are the expectations for users there?

Also it is now my expectation that xla and ipu APIs are being used already? And we'd probably only want to move forward with a complete removal of the xla/ipu ones if @cowanmeg @charlie-wt and others involved sign off on it.

So TLDR, I think its worth stepping back a bit from this PR for now, as for the checkpoint regression with mps, that should just be a simple fix we can do on the side. Of course if you are interested in writing up a higher-level proposal of what amp's API surface should look like, we'd be happy to discuss that in an issue.

**Comment by heidongxianhua** [comment]
thanks for your reply. And as for custom backends, I think it will be better to call a function with `device` argument rather than to to call the funcs defined in device modules like `torch.cuda`. And  the best thing is to deprecate the specific APIs for device, and use the unified api.

**Comment by charlie-wt** [comment]
from an ipu perspective, the `*_autocast_ipu_*` api isn't being used directly from our out-of-tree backend and there's not yet a release which includes the whole autocast support, so i don't have an issue with changing these functions to something more device-agnostic.

**Comment by cowanmeg** [comment]
We are in a similar situation for xla. Since we just added support we don't have many users, so no issues changing these functions. We are also not using set_autocast_xla_dtype or get_autocast_xla_dtype in our codebase.

**Comment by soulitzer** [comment]
Thanks for the status updates! Though as a heads up, I don't plan on looking into a refactor of the API surface of amp too much in the near term so unless @heidongxianhua or someone else decides take that on, we may need to ping yall again later just to double check)when we pick this up again.

**Comment by heidongxianhua** [comment]
> Thanks for the status updates! Though as a heads up, I don't plan on looking into a refactor of the API surface of amp too much in the near term so unless @heidongxianhua or someone else decides take that on, we may need to ping yall again later just to double check)when we pick this up again.

yeah, as we discussed here https://github.com/pytorch/pytorch/pull/104140#discussion_r1254088506  we are working on and want refactor the device  specific APIs , and use the same and unifed APIs to do same things for different devices with `device` argument. And I think it will be fancy to support different devices and it is easy to add a new backend without change the surface APIs. @soulitzer

**Comment by soulitzer** [comment]
> yeah, as we discussed here https://github.com/pytorch/pytorch/pull/104140#discussion_r1254088506 we are working on and want refactor the device specific APIs , and use the same and unifed APIs to do same things for different devices with device argument. And I think it will be fancy to support different devices and it is easy to add a new backend without change the surface APIs

@heidongxianhua thanks for the context there.

> For example, we have both torch.cuda.amp.autocast(args...) and torch.autocast("cuda", args...).
We also have a bunch of torch.autocast_set_enabled APIs directly on the torch name space, when we really want them to live in the amp namespace. They aren't really documented anywhere either. How do we expect users to use those low-level APIs or should they be private now?
We also have the issue of custom backends. What are the expectations for users there?

I still think some of the questions here are worth answering, namely in an ideal world what do we envision AMP's public API surface to look like (1) for devices like cuda and cpu (2) custom devices. 
For this PR we are introducing new public APIs to the torch name space (not easily reversible!), e.g., if we eventually actually prefer autocast apis to live in torch.amp.*, then maybe this is not the right direction. Or do we expect users to mainly rely on the context manager, and not use these lower level get/set APIs?
I think its worth creating a separate issue where these things can be discussed with more folks.
